### PR TITLE
FW-517 & FW-536: private directive issues

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/entry_point/src/generator.ts [hunk 1]
```diff
@@ -28,7 +29,7 @@ export class FlatIndexGenerator implements ShimGenerator {
     // If there's only one .ts file in the program, it's the entry. Otherwise, look for the shortest
     // (in terms of characters in the filename) file that ends in /index.ts. The second behavior is
     // deprecated; users should always explicitly specify a single .ts entrypoint.
-    const tsFiles = files.filter(isNonDeclarationTsFile);
+    const tsFiles = files.filter(file => !file.endsWith('.d.ts'));
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/entry_point/src/generator.ts]: Why de-abstract this?

Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/entry_point/src/generator.ts]: Fixed :)

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 2]
```diff
@@ -1136,40 +1138,83 @@ describe('ngtsc behavioral tests', () => {
...
+
+        @Component({
+          template: '<div test></div>',
+        })
+        class Cmp {}
+
+        @NgModule({
+          declarations: [Cmp],
+          // Multiple imports of the same module used to result in duplicate directive references
+          // in the output.
```
Comment (gkalpak) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: This comment is irrelevant here.

File: packages/compiler-cli/src/ngcc/src/packages/transformer.ts [hunk 3]
```diff
@@ -103,11 +103,12 @@ export class Transformer {
...
     const referencesRegistry = new NgccReferencesRegistry(reflectionHost);
     const decorationAnalyzer = new DecorationAnalyzer(
-        typeChecker, reflectionHost, referencesRegistry, bundle.rootDirs, isCore);
+        bundle.src.program, bundle.src.options, bundle.src.host, typeChecker, reflectionHost,
+        referencesRegistry, bundle.rootDirs, isCore);
     const switchMarkerAnalyzer = new SwitchMarkerAnalyzer(reflectionHost);
     const privateDeclarationsAnalyzer =
         new PrivateDeclarationsAnalyzer(reflectionHost, referencesRegistry);
-    const decorationAnalyses = decorationAnalyzer.analyzeProgram(bundle.src.program);
+    const decorationAnalyses = decorationAnalyzer.analyzeProgram();
```
Comment (AndrewKushnir) [inline packages/compiler-cli/src/ngcc/src/packages/transformer.ts]: This call looks inconsistent with other Analyzers. Just curious if it's possible to maintain the contract (sorry if I missed it in commit descriptions).

**Review by AndrewKushnir — APPROVED** [review]
Looks great, thanks @alxhub.

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Fixed!

File: packages/compiler-cli/src/ngtsc/imports/src/resolver.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import {isFromDtsFile} from '../../util/src/typescript';
+
+import {AbsoluteReference, Reference, ResolvedReference} from './references';
+
+export interface ReferenceResolver {
+  resolve(decl: ts.Declaration, importFromHint: string|null, fromFile: string):
+      Reference<ts.Declaration>;
+}
+
+export class NpmReferenceResolver implements ReferenceResolver {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/imports/src/resolver.ts]: A better name would be the `TsReferenceResolver` - renamed.

**Comment by alxhub** [comment]
@AndrewKushnir you're correct in that it makes this particular "Analyzer" function a bit differently than the others. I actually think this version is correct. The other analyzers have a signature like:

```typescript
class Analyzer {
  constructor(private checker: ts.TypeChecker) {}
  analyzeProgram(program: ts.Program) {}
}
```

This is arguably incorrect. A `ts.TypeChecker` is created from a `ts.Program` - it makes no sense to analyze one program with the `TypeChecker` from another. So even though this API _looks like_ a single instance of the analyzer can run over multiple programs, this would not work in practice, and `analyzeProgram` *must* be called with the same `Program` as was used to get the `TypeChecker` the analyzer was constructed with.

Thus, I think my refactor actually makes the API better, and the other analyzer should follow suit.

File: packages/compiler-cli/src/ngtsc/entry_point/src/reference_graph.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    }
+    // Consider outgoing edges from `source`.
+    seen.add(source);
+
+    if (!this.references.has(source)) {
+      // There are no outgoing edges from `source`.
+      return null;
+    } else {
+      // Look through the outgoing edges of `source`.
+      // TODO(alxhub): use proper iteration when build.sh is removed. (#27762)
```
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/src/ngtsc/entry_point/src/reference_graph.ts]: `build.sh` has been removed in #27937 fwiw :wink:

[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/entry_point/src/reference_graph.ts]: 🎉🍾🎆

I'll go through and convert back to beautiful native iteration.

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 6]
```diff
@@ -1193,5 +1195,112 @@ describe('ngtsc behavioral tests', () => {
...
+      env.write('test.ts', `
+        import {Directive, NgModule} from '@angular/core';
+
+        // The directive is not exported.
+        @Directive({selector: 'test'})
+        class Dir {}
+
+        // Neither is the module which declares it - meaning the directive is not visible here.
+        @NgModule({declarations: [Dir], exports: [Dir]})
+        class DirModule {}
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts]: Added!

File: packages/compiler-cli/src/ngtsc/entry_point/src/reference_graph.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/entry_point/src/reference_graph.ts]: Noooooo!!

The legacy build on CircleCI is basically build.sh reincarnated. It has the same issue 😭

**Review by kara — APPROVED** [review]
LGTM

**Review by gkalpak — APPROVED** [review]
LGTM (with a couple of minor comments)

**Comment by alxhub** [comment]
@gkalpak 

> I'm mentioning it because there is an i here, but there is no i is several other places and I don't expect it to make any difference (and it is indeed existing code being moved around), so 👍

You're correct, and this should be unified, but it's outside the scope of this PR for now.

> Then why doesn't it have `expressable = false`? 😕

It should - I'll fix this in a bug fix PR.

> This is marked as resolved, but I still see this unexpectedly formatted 😕
I would expect `ResolvedValue` to be on the same line as the preceeding `->`

As would I :( unfortunately the auto-formatter disagrees - it changes it back. I don't want to disable it just for this comment.

**Review by IgorMinar — APPROVED** [review]
LGTM

@alxhub one nit: I'd prefer the `feat(forms)` commit message to have a more customer friendly subject that we'll end up in the changelog (e.g. list the name of main directives we are exporting)

**Comment by alxhub** [comment]
Dear karataker,

The bot still thinks I need two approvals, but @IgorMinar already approved and should count for fw-public-api, and I'm the approver for fw-compiler.

**Pull request closed without merging**